narwhal/primary/src/certificate_waiter.rs

```diff
@@ -3,226 +3,380 @@
 // SPDX-License-Identifier: Apache-2.0
 use crate::metrics::PrimaryMetrics;
 use config::Committee;
-use dashmap::DashMap;
-use futures::future::try_join_all;
-use once_cell::sync::OnceCell;
-use std::sync::Arc;
+use crypto::{NetworkPublicKey, PublicKey};
+use fastcrypto::Hash;
+use futures::{future::try_join_all, stream::FuturesUnordered, StreamExt};
+use itertools::Itertools;
+use network::PrimaryToPrimaryRpc;
+use rand::{rngs::ThreadRng, seq::SliceRandom, Rng};
+use std::{collections::BTreeMap, sync::Arc, time::Duration};
 use storage::CertificateStore;
 use tokio::{
-    sync::{oneshot, watch},
+    sync::{mpsc, watch},
     task::JoinHandle,
-    time::{sleep, Duration, Instant},
+    time::{self, Instant},
 };
-use tracing::info;
+use tracing::{debug, error, instrument, trace, warn};
 use types::{
-    bounded_future_queue::BoundedFuturesUnordered,
     error::{DagError, DagResult},
     metered_channel::{Receiver, Sender},
-    try_fut_and_permit, Certificate, CertificateDigest, HeaderDigest, ReconfigureNotification,
+    Certificate, FetchCertificatesRequest, FetchCertificatesResponse, ReconfigureNotification,
     Round,
 };
 
 #[cfg(test)]
 #[path = "tests/certificate_waiter_tests.rs"]
 pub mod certificate_waiter_tests;
 
-/// The resolution of the timer that checks whether we received replies to our parent requests, and triggers
-/// a round of GC if we didn't.
-const GC_RESOLUTION: u64 = 10_000;
+// Maximum number of certficates to fetch with one request.
+const MAX_CERTIFICATES_TO_FETCH: usize = 1000;
 
-/// Waits to receive all the ancestors of a certificate before looping it back to the `Core`
-/// for further processing.
+/// When there are certificates missing from local store, e.g. discovered when a received certificate has missing parents,
+/// CertificateWaiter is responsible for fetching missing certificates from other primaries.
 pub struct CertificateWaiter {
+    /// Identity of the current authority.
+    name: PublicKey,
     /// The committee information.
     committee: Committee,
+    /// Network client to fetch certificates from other primaries.
+    network: Arc<dyn PrimaryToPrimaryRpc>,
```

Comment:
I want to be able to supply a fake implementation of `PrimaryToPrimaryRpc` for testing. I can look into using a generic instead.

Reply:
You can also use a real network in tests with a mock service. Take a look at `block_remover_tests.rs` use of `MockPrimaryToWorker` for an example of how this looks. I added generated mocks for all of the narwhal services to our code gen.

Reply:
If it's possible for you to follow the pattern other components have of using the `P2pNetwork` directly and then the mock servers in tests, it'll make it easier for me later because I think we want to clean up `P2pNetwork` and remove all these networking RPC traits entirely (at least the ones that are not anemo generated)

Reply:
@aschran thanks for the example in `block_remover_tests.rs`. My only comment is that sometimes we only want to use a fake / mock to test the client interactions. Here we had to create fake / mock servers to achieve that and go through real network. I hope faking client can be simplified in future.

Reply:
Agreed using the real network isn't ideal. I think we want to refactor away P2pNetwork and can hopefully improve the testing story all at once as part of that work. @bmwill in case you have any thoughts to add?

Reply:
@aschran that will be really cool. Look forward to that.